include/corpus/metadata.h

```diff
@@ -0,0 +1,317 @@
+/**
+ * @file metadata.h
+ * @author Chase Geigle
+ *
+ * All files in META are dual-licensed under the MIT and NCSA licenses. For more
+ * details, consult the file LICENSE.mit and LICENSE.ncsa in the root of the
+ * project.
+ */
+
+#ifndef META_CORPUS_METADATA_H_
+#define META_CORPUS_METADATA_H_
+
+#include <cstdint>
+#include <vector>
+#include <string>
+
+#include "cpptoml.h"
+#include "io/binary.h"
+#include "util/optional.h"
+
+namespace meta
+{
+namespace corpus
+{
+
+/**
+ * Represents the collection of metadata for a document.
+ */
+class metadata
+{
+  public:
+    /**
+     * Type tag for a field.
+     */
+    enum class field_type : uint8_t
+    {
+        SIGNED_INT = 0,
+        UNSIGNED_INT,
+        DOUBLE,
+        STRING
+    };
+
+    /**
+     * Pair for storing the schema: contains its name and type.
+     */
+    struct field_info
+    {
+        std::string name;
+        field_type type;
+
+        field_info() = default;
+        field_info(std::string n, field_type ft) : name{std::move(n)}, type{ft}
+        {
+            // nothing
+        }
+        field_info(const field_info&) = default;
+        field_info(field_info&&) = default;
+        field_info& operator=(const field_info&) = default;
+        field_info& operator=(field_info&&) = default;
+        ~field_info() = default;
+    };
+
+    // I want the below to be a const field_info, but g++ gives a cryptic
+    // compiler error in that case... clang++ accepts it just fine. -sigh-
+    using schema = std::vector<field_info>;
+
+    metadata(const char* start, const schema& sch)
+        : schema_{sch}, stream_{start}
+    {
+        // nothing
+    }
+
+    /**
+     * @param name The metadata field to obtain
+     * @return the metadata associated with that field, if it exists,
+     * converted to type T.
+     */
+    template <class T>
+    util::optional<T> get(const std::string& name)
+    {
+        for (uint64_t i = 0; i < schema_.size(); ++i)
+        {
+            switch (schema_[i].type)
+            {
+                case field_type::SIGNED_INT:
+                    int64_t si;
```

Comment:
I actually don't know.

I think it's fine for integers, since there isn't any initialization that's being skipped, but I'm not sure why it doesn't complain later down when I construct a string, though... is it maybe because it's the last case and can't be jumped over? In either case, I can put braces around the cases to be more explicit if you want.